### PR TITLE
[None][fix] Restore DSv4 NVFP4 routed swiglu_limit on TRTLLM-Gen

### PR DESCRIPTION
**Pull request by Barry-Delaney**
## Description

Cherry-picks #14673 onto main after the original fix landed on feat/deepseek_v4 but was not carried over with the model split.

- Restore the routed-expert swiglu_limit for TRTLLM-Gen NVFP4.
- Move cubin-domain bias and clamp normalization into NVFP4TRTLLMGenFusedMoEBaseMethod so the plain DeepSeek-V4 SwiGLU path applies it.
- Preserve the current FP8 scalar-only handling and account for the removal of WideEP.

## Testing

- Pre-merge CI

## Dev Engineer Review

- Restores per-expert `swiglu_limit` for TRTLLM-Gen NVFP4.
- Applies bias and SwiGLU parameter normalization in the shared base method.
- Preserves scalar-only handling for TRTLLM-Gen FP8 block-scale MoE.
- Avoids scaling synthetic zero-bias tensors.
- No configuration or test-list files changed.
- No clear API or performance regression found.

## QA Engineer Review

- Added tests for:
  - NVFP4 per-expert `swiglu_limit` preservation.
  - Expansion for local EPLB slots.
  - Scalar-only limits for TRTLLM-Gen FP8 block-scale MoE.
- The tests are in `tests/unittest/_torch/modeling/test_modeling_deepseekv4.py`.
- No corresponding `test-db/` or `qa/` coverage entry was identified.
- **Verdict: needs follow-up.**

### CODE REVIEW

**Review by sunnyqgg — APPROVED** [review]
LGTM

**Review by leslie-fang25 — APPROVED** [review]

File: tensorrt_llm/_torch/moe/fused_moe/quantization.py [hunk 1]
```diff
@@ -5244,6 +5244,17 @@ def process_weights_after_loading(self,
         # Finalize shared expert alphas and fc31_scale_c for online EPLB
         self._finalize_shared_expert_alphas(module)
 
+        # Cubin clamp / GLU bias inputs are consumed in the pre-dequant GEMM
+        # output domain (i.e. divided by fc31_alpha / fc2_alpha).
+        if getattr(module, 'w3_w1_bias', None) is not None:
+            module.w3_w1_bias.data.div_((module.fc31_alpha.data).view(-1, 1))
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/moe/fused_moe/quantization.py:5259]: Moving this into the base method also makes it run for `W4A8NVFP4FP8TRTLLMGenFusedMoEMethod`, which previously never applied it. On that path `w3_w1_bias`/`w2_bias` are auto-created as zeros (`fused_moe_trtllm_gen.py:673`) and `_check_configs` rejects a real bias, so it's a mathematical no-op — except that `fc31_alpha = w_scale_2 / input_scale` is zero for an all-zero expert slot (padded/unused slots under EPLB), and `0 / 0` gives NaN bias rather than the zeros the kernel expects. Worth either skipping the division when the tensor is known-zero or clamping alpha, and stating in the comment that the widened applicability is intentional.

File: tensorrt_llm/_torch/models/modeling_deepseekv4.py [hunk 2]
```diff
@@ -1566,24 +1566,13 @@ def __init__(
...
             requires_scalar_only_swiglu_limit = (
                 moe_cls is TRTLLMGenFusedMoE
                 and experts_quant_config.quant_mode.has_fp8_block_scales()
             )
-            if (
-                supports_swiglu_limit
-                and not kernel_requires_bias_for_swiglu_limit
-                and not requires_scalar_only_swiglu_limit
-            ):
+            if supports_swiglu_limit and not requires_scalar_only_swiglu_limit:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_deepseekv4.py:1580]: Nothing pins the restored behavior. This regressed exactly because a model split dropped the fix, and there's no test asserting that a DSv4 NVFP4 + TRTLLM-Gen layer receives a per-expert `swiglu_limit` tensor (and the FP8 block-scale layer only the scalar). A small unit test around this resolver block — no GPU needed if it only inspects the value passed to `create_moe` — would stop the next split from silently dropping it again.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

The mechanism checks out: `TRTLLMGenFusedMoE._get_quant_method` returns `NVFP4TRTLLMGenFusedMoEBaseMethod` for plain SwiGLU NVFP4 (no `swiglu_alpha`, not SiTu, activation not Relu2/Silu) — that's the DeepSeek-V4 routed path — and the pre-dequant-domain division lived only in the padded `NVFP4TRTLLMGenFusedMoEMethod` subclass, so the limit reached the cubin in the wrong domain. Moving it to the base is the right place.

Two asks:

- The guard being removed was justified by measured near-zero accuracy. Please attach before/after accuracy numbers for DSv4 NVFP4 on TRTLLM-Gen (the config that motivated the original guard), not just "pre-merge CI" — CI has no test that exercises this combination. Also, the release notes attached to the PR say pre-merge CI failed; worth confirming that's resolved or unrelated before merge.
- This is a fix for a real accuracy regression that a model split silently dropped; it should carry an NVBug in the title tag rather than `[None]`, so the same loss is traceable next time.

**Comment by xxi-nv** [comment]
https://github.com/NVIDIA/TensorRT-LLM/pull/18409 depends on this PR.

File: tensorrt_llm/_torch/moe/fused_moe/quantization.py (hunk 1 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/moe/fused_moe/quantization.py:5259]: Checked this against the PR head (`8c8f0c1`). The factual chain holds, but the specific NaN trigger does not, and I think a different fix is the right one.

**Confirmed**

- The widened applicability is real: `W4A8NVFP4FP8TRTLLMGenFusedMoEMethod` (`quantization.py:5752`) inherits the same base and its `process_weights_after_loading` (`:5787-5788`) is a bare `return super().…(num_elts_per_sf=32)`, so it now runs this block. The deleted override lived on the sibling `NVFP4TRTLLMGenFusedMoEMethod` (`:5360`).
- `fused_moe_trtllm_gen.py:671-682` does auto-create zero `w3_w1_bias`/`w2_bias` for `has_w4a8_nvfp4_fp8` when `not self.bias`.
- A real bias is indeed rejected: the `_check_configs` allowlist (`fused_moe_trtllm_gen.py:555-557`) omits `has_w4a8_nvfp4_fp8`, and `QuantAlgo.W4A8_NVFP4_FP8` maps to a QuantMode that sets only the `W4A8_NVFP4_FP8` bit (`mode.py:434-435`), so `has_nvfp4()` (`mode.py:188`) is `False`.
- The arithmetic is right: `load_expert_fc31_alpha_nvfp4` (`quantization.py:2352`) yields `alpha == 0` when `weight_scale_2 == 0`, and `0 / 0` is NaN.

**Does not hold: the EPLB padded/unused slot case**

- `_reconcile_and_compute_alphas` skips slots whose scales were never loaded (`quantization.py:2777`), and `fc31_alpha`/`fc2_alpha` are initialized to `torch.ones`, not zeros (`:2290`, `:2295`). An unloaded slot therefore gets `0 / 1 == 0` — the intended no-op, not NaN.
- EPLB has no unused slots: `interface.py:552-555` asserts `len(initial_global_assignments) == num_slots` **and** `set(initial_global_assignments) == set(range(num_experts))`, so every slot — including the extra ones when `num_slots > num_experts` — maps to a real expert id and loads real weights.
- So `alpha == 0` would require a checkpoint that actually contains an all-zero (dead) expert. Possible, but unrelated to EPLB.

**And it would not corrupt output even then**

The `has_w4a8_nvfp4_fp8` forward branch (`fused_moe_trtllm_gen.py:1049-1080`, `run_moe` → `fp8_fp4_block_scale_moe_runner`) does not take `w3_w1_bias` / `w2_bias` / `swiglu_*` at all. On this path those tensors are never consumed by the kernel, so a NaN in them cannot reach the output.

**But the block should still be gated, for a more direct reason**

The base method is now internally inconsistent: `_shuffle_all_experts` (`quantization.py:5304-5308`) gates bias handling on `if module.bias:`, while the new block gates on attribute existence. On the NVFP4 path the two are equivalent (no auto-created bias); on the w4a8_nvfp4_fp8 path they are not, so the auto-created zero bias is **not shuffled but is divided**. Gating the division on `module.bias` restores that consistency and removes the NaN question entirely, with no alpha clamp needed:

```python
        # Cubin clamp / GLU bias inputs are consumed in the pre-dequant GEMM
        # output domain (i.e. divided by fc31_alpha / fc2_alpha).
        #
        # Gated on module.bias to match _shuffle_all_experts above. Moving this
        # into the base method deliberately widens it to
        # W4A8NVFP4FP8TRTLLMGenFusedMoEMethod, whose kernel
        # (fp8_fp4_block_scale_moe_runner) takes no bias: that path
        # auto-creates zero bias tensors (fused_moe_trtllm_gen.py:671-682) that
        # must not be divided, since a zero alpha from an all-zero expert slot
        # would turn them into NaN.
        if module.bias:
            module.w3_w1_bias.data.div_((module.fc31_alpha.data).view(-1, 1))
            module.w2_bias.data.div_((module.fc2_alpha.data).view(-1, 1))
        if getattr(module, 'swiglu_beta', None) is not None:
            module.swiglu_beta.data.div_((module.fc31_alpha.data))
        if getattr(module, 'swiglu_limit', None) is not None:
            module.swiglu_limit.data.div_((module.fc31_alpha.data))
```

`swiglu_beta` / `swiglu_limit` keep the `getattr` form on purpose — they are created independently of `module.bias` (e.g. the SiTu path at `fused_moe_trtllm_gen.py:655-665`), and neither can exist on the w4a8_nvfp4_fp8 path because the same `_check_configs` gate rejects them there.

File: tensorrt_llm/_torch/models/modeling_deepseekv4.py (hunk 2 above)
Comment (xxi-nv) [reply tensorrt_llm/_torch/models/modeling_deepseekv4.py:1580]: Agreed, and it is cheap to add. Current state: `swiglu_limit` appears exactly twice in `tests/unittest/_torch/modeling/test_modeling_deepseekv4.py` (`:132`, `:140`), both inside `test_deepseek_v4_config_aliases` asserting the config alias. Nothing asserts what the resolver hands to `create_moe`.

There is already a working no-GPU template for this exact shape of assertion: `tests/unittest/_torch/models/test_qwen3_next_moe_quant.py:210-238` builds a `SimpleNamespace` model config, declares the hardware with `override_moe_environment(MoEEnvironment(sm=100))`, patches the resolver, and aborts `__init__` with a sentinel exception. For DSv4 it is simpler still — patch `create_moe` itself and inspect the kwargs. `moe_resolution.py` has no direct `get_sm_version`/CUDA dependency, and the resolver block already does `device = "cuda" if torch.cuda.is_available() else "cpu"` (`modeling_deepseekv4.py:1583`), so a CPU-only run works.

Sketch:

```python
class _StopInit(Exception):
    pass


def _captured_swiglu_limit(quant_algo, moe_backend="TRTLLM"):
    captured = {}

    def _capture(**kwargs):
        captured["swiglu_limit"] = kwargs.get("swiglu_limit")
        captured["swiglu_limit_scalar"] = kwargs.get("swiglu_limit_scalar")
        raise _StopInit

    with (
        override_moe_environment(MoEEnvironment(sm=100)),
        patch("tensorrt_llm._torch.models.modeling_deepseekv4.create_moe", _capture),
        pytest.raises(_StopInit),
    ):
        # model_config: DeepseekV4Config with swiglu_limit=SWIGLU_LIMIT,
        # QuantConfig(quant_algo=quant_algo), moe_backend=moe_backend
        DeepseekV4MoE(model_config, ...)

    return captured


def test_nvfp4_trtllm_gen_gets_per_expert_swiglu_limit():
    captured = _captured_swiglu_limit(QuantAlgo.NVFP4)
    # This is the assertion the model split would have failed.
    assert captured["swiglu_limit"] is not None
    assert captured["swiglu_limit"].numel() == LOCAL_NUM_SLOTS
    assert captured["swiglu_limit_scalar"] == pytest.approx(SWIGLU_LIMIT)


def test_fp8_block_scale_trtllm_gen_gets_scalar_only():
    captured = _captured_swiglu_limit(QuantAlgo.FP8_BLOCK_SCALES)
    assert captured["swiglu_limit"] is None
    assert captured["swiglu_limit_scalar"] == pytest.approx(SWIGLU_LIMIT)
```

Two cases cover both halves of the resolver: NVFP4 must get the per-expert tensor (the behavior this PR restores), and FP8 block-scale must get the scalar only (`requires_scalar_only_swiglu_limit`). A third case with an unsupported backend asserting both are `None` would pin the `supports_swiglu_limit` branch too.

**Comment by xxi-nv** [comment]
CI is broken; to save the CI resource, I will submit this commit in https://github.com/NVIDIA/TensorRT-LLM/pull/18409
